### PR TITLE
Automation test 

### PR DESCRIPTION
**Pull request by labkey-sweta**
#### Rationale
Automation test coverage for Issue 44511: Ability to change a field's data type to known safe alternatives

#### Related Pull Requests
* 

#### Changes
New test class src/org/labkey/test/tests/DomainFieldTypeChangeTest.java

### CODE REVIEW

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        domainFormPanel.getField("testInteger").setType(FieldDefinition.ColumnType.Decimal);
+        domainFormPanel.getField("testBoolean").setNumberFormat("yes;no");
+        domainDesignerPage.clickFinish();
+
+        clickAndWait(Locator.linkWithText(listName));
+        DataRegionTable table = new DataRegionTable("query", getDriver());
+        checker().verifyEquals("Incorrect values after changing integer to decimal", Arrays.asList("1.0", "2.0", "3.0"),
+                table.getColumnDataAsText("testInteger"));
+
+        log("Verifying changin data fields to string");
```
Comment (labkey-jeckels) [inline src/org/labkey/test/tests/DomainFieldTypeChangeTest.java]:
```suggestion
        log("Verifying changing data fields to string");
```

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+        clickAndWait(Locator.linkWithText(listName));
+        table = new DataRegionTable("query", getDriver());
+        checker().verifyEquals("Incorrect values after changing integer to string", Arrays.asList("1", "2", "3"),
+                table.getColumnDataAsText("testInteger"));
+        checker().verifyEquals("Incorrect values after changing decimal to string", Arrays.asList("1.1", "2.2", "3.3"),
+                table.getColumnDataAsText("testDecimal"));
+        checker().verifyEquals("Incorrect values after changing date to string", Arrays.asList("2022-01-01 00:00:00", "2022-01-02 00:00:00", "2022-01-03 00:00:00"),
+                table.getColumnDataAsText("testDate"));
+        checker().verifyEquals("Incorrect values after changing boolean to string", Arrays.asList("yes", "no", "yes"),
```
Comment (labkey-jeckels) [inline src/org/labkey/test/tests/DomainFieldTypeChangeTest.java]: To confirm that the tyoe migration was successful, I recommend also doing an update of one of these fields to enter a value that wasn't previously supported, like "1Test" for testInteger.

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        domainFormPanel.getField("testBoolean").setType(FieldDefinition.ColumnType.String);
+        domainDesignerPage.clickFinish();
+
+        clickAndWait(Locator.linkWithText(listName));
+        table = new DataRegionTable("query", getDriver());
+        checker().verifyEquals("Incorrect values after changing integer to string", Arrays.asList("1", "2", "3"),
+                table.getColumnDataAsText("testInteger"));
+        checker().verifyEquals("Incorrect values after changing decimal to string", Arrays.asList("1.1", "2.2", "3.3"),
+                table.getColumnDataAsText("testDecimal"));
+        checker().verifyEquals("Incorrect values after changing date to string", Arrays.asList("2022-01-01 00:00:00", "2022-01-02 00:00:00", "2022-01-03 00:00:00"),
```
Comment (labkey-jeckels) [inline src/org/labkey/test/tests/DomainFieldTypeChangeTest.java]: Does this work on both Postgres and SQLServer? I believe that they render date->string values differently.

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        clickButton("Save and Finish");
+
+        goToManageAssays();
+        clickAndWait(Locator.linkWithText(assayName));
+        waitForElement(Locator.linkWithText(runName));
+        click(Locator.linkWithText("Manage assay design"));
+        assayDesignerPage = _assayHelper.clickEditAssayDesign();
+        runFields = assayDesignerPage.goToRunFields();
+        runFields.getField("runTestInteger").setType(FieldDefinition.ColumnType.Decimal);
+        checker().verifyEquals("Incorrect error message for integer to decimal", Arrays.asList("Cannot convert from INTEGER to DOUBLE for non-provisioned table"), assayDesignerPage.clickSaveExpectingErrors());
```
Comment (labkey-jeckels) [inline src/org/labkey/test/tests/DomainFieldTypeChangeTest.java]: This should be supported. I can see the problem in the code. Can you log a bug please?

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    @Override
+    protected void doCleanup(boolean afterTest) throws TestTimeoutException
+    {
+        super.doCleanup(afterTest);
+    }
+
+    @Override
+    public List<String> getAssociatedModules()
+    {
+        return null;
```
Comment (labkey-jeckels) [inline src/org/labkey/test/tests/DomainFieldTypeChangeTest.java:217]: Is this correct? Should we return anything here?

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        checker().verifyEquals("Batch fields : Incorrect value after changing Date to string", Arrays.asList("2022-01-01 00:00:00"),
+                table.getColumnDataAsText("Batch/batchTestDate"));
+        checker().verifyEquals("Batch fields : Incorrect value after changing Decimal to string", Arrays.asList("1.1"),
+                table.getColumnDataAsText("Batch/batchTestDecimal"));
+
+        checker().screenShotIfNewError("AfterRunAndBatchChanges");
+    }
+
+    @Override
+    protected void doCleanup(boolean afterTest) throws TestTimeoutException
```
Comment (labkey-jeckels) [inline src/org/labkey/test/tests/DomainFieldTypeChangeTest.java]: This override isn't doing anything useful - can it be deleted?

**Review by labkey-jeckels — APPROVED** [review]

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java (hunk 5 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/DomainFieldTypeChangeTest.java:217]: Thats a legacy method, should be present.

File: src/org/labkey/test/tests/DomainFieldTypeChangeTest.java (hunk 3 above)
Comment (labkey-sweta) [reply src/org/labkey/test/tests/DomainFieldTypeChangeTest.java]: Changed the code accordingly

**Pull request merged**